File: core/core-domain/src/main/java/com/nmh/commerce/coupon/IssuedCoupon.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.user.User;
import lombok.Builder;
import lombok.Getter;

import java.time.LocalDateTime;

@Getter
public class IssuedCoupon {
    private final Long id;
    private final Coupon coupon;
    private final User user;
    private final LocalDateTime issuedAt;

    @Builder
    private IssuedCoupon(Long id, Coupon coupon, User user, LocalDateTime issuedAt) {
        this.id = id;
        this.coupon = coupon;
        this.user = user;
        this.issuedAt = issuedAt;
    }

    public static IssuedCoupon issue(Coupon coupon, User user, LocalDateTime issuedAt) {
        return IssuedCoupon.builder()
            .coupon(coupon)
            .user(user)
            .issuedAt(issuedAt)
            .build();
    }

}

File: core/core-domain/src/main/kotlin/com/nmh/commerce/coupon/IssuedCoupon.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.user.User
import java.time.LocalDateTime

data class IssuedCoupon(
    val id: Long = 0,
    val coupon: Coupon,
    val user: User,
    val issuedAt: LocalDateTime,
) {
    companion object {
        fun issue(
            coupon: Coupon,
            user: User,
            issuedAt: LocalDateTime,
        ): IssuedCoupon = IssuedCoupon(coupon = coupon, user = user, issuedAt = issuedAt)
    }
}
